src/launchpad/kafka.py

```diff
@@ -0,0 +1,327 @@
+                    if asyncio.iscoroutine(result):
+                        logger.warning("Message handler returned a coroutine but we're in sync context - ignoring")
+                except Exception as e:
+                    logger.error(f"Error in message handler: {e}", exc_info=True)
+            else:
+                logger.info(f"No handler set, received message: {launchpad_msg}")
+                payload = launchpad_msg.get_json_payload()
+                logger.info(f"Message payload: {payload}")
+
+        except Exception as e:
+            logger.error(f"Error processing message: {e}", exc_info=True)
+
+    def close(self) -> None:
+        pass
+
+    def terminate(self) -> None:
+        pass
+
+    def join(self, timeout: Optional[float] = None) -> None:
+        pass
+
+
+class MessageProcessingStrategyFactory(ProcessingStrategyFactory[KafkaPayload]):
+    """Factory for creating message processing strategies."""
+
+    def __init__(self, message_handler: Optional[Callable[[LaunchpadMessage], Any]] = None) -> None:
+        self.message_handler = message_handler
+
+    def create_with_partitions(
+        self,
+        commit: Commit,
+        partitions: Mapping[Partition, int],
+    ) -> ProcessingStrategy[KafkaPayload]:
+        """Create a processing strategy with partition information."""
+        # Use the message processing strategy directly
+        return MessageProcessingStrategy(self.message_handler)
+
+
+class KafkaConsumer:
+    """Arroyo-based Kafka consumer for Launchpad."""
+
+    def __init__(
+        self,
+        topics: List[str],
+        group_id: str,
+        bootstrap_servers: Optional[str] = None,
+        message_handler: Optional[Callable[[LaunchpadMessage], Any]] = None,
+    ) -> None:
+        self.topics = topics
+        self.group_id = group_id
+        # Ensure bootstrap_servers is always a string
+        self.bootstrap_servers: str = bootstrap_servers or os.getenv(
+            "KAFKA_BOOTSTRAP_SERVERS"
+        )  # type: ignore[assignment]
+        if not self.bootstrap_servers:
+            raise ValueError("KAFKA_BOOTSTRAP_SERVERS environment variable must be set")
+        self.message_handler = message_handler
+        self._processor: Optional[StreamProcessor[KafkaPayload]] = None
+        self._shutdown_requested = False
+
+        logger.info(f"Initialized Arroyo Kafka consumer for topics: {topics}")
+        logger.info(f"Group ID: {group_id}")
+        logger.info(f"Bootstrap servers: {self.bootstrap_servers}")
+
+    def run(self) -> None:
+        """Run the Kafka consumer (blocking, like Snuba)."""
+        # Ensure required topics exist before starting consumer
+        try:
+            ensure_topics_exist(
+                bootstrap_servers=self.bootstrap_servers,
+                topic_names=self.topics,
+                num_partitions=1,
+                replication_factor=1,
+            )
```

Comment:
When you get closer to production, you'll need a way to disable this logic as topics, partitions and kafka clusters are configured through getsentry/ops. The streaming teams can help with getting your topic setup.